Mathlib/Data/Multiset/Order.lean

```diff
@@ -0,0 +1,552 @@
+
+## References
+
+* [Wikipedia, Dershowitz–Manna ordering*]
+(https://en.wikipedia.org/wiki/Dershowitz%E2%80%93Manna_ordering)
+
+* [CoLoR](https://github.com/fblanqui/color), a Coq library on rewriting theory and termination.
+  Our code here is inspired by their version of called `mOrd_wf` in the file
+  [MultisetList.v](https://github.com/fblanqui/color/blob/1.8.5/Util/Multiset/MultisetList.v).
+
+-/
+
+variable {α : Type*}
+
+/-- The standard Dershowitz–Manna ordering: -/
+inductive MultisetLT [DecidableEq α] [Preorder α] (M N : Multiset α) : Prop :=
+  | MLT : ∀ (X Y Z: Multiset α),
+    Y ≠ ∅ →
+    M = Z + X →
+    N = Z + Y →
+    (∀ x ∈ X, ∃ y ∈ Y, x < y) → MultisetLT M N
+
+/-- MultisetRedLt is a special case of MultisetLT. The transitive closure of it is used to define
+an equivalent (proved later) version of the ordering. -/
+inductive MultisetRedLt [DecidableEq α][LT α] (M N : Multiset α) : Prop :=
+  | RedLt : ∀ (X Y:Multiset α) (a : α) ,
+    (M = X + Y) →
+    (N = X + {a}) →
+    (∀ y, y ∈ Y → y < a) → MultisetRedLt M N
+
+open Relation
+
+/-- MultisetLt is the transitive closure of MultisetRedLt and is equivalent to MultisetLT
+    (proved later). -/
+def MultisetLt [DecidableEq α][LT α] : Multiset α → Multiset α → Prop := TransGen MultisetRedLt
+
+/-- AccM_1 defines the accessibility relation given MultisetRedLt. -/
+def AccM_1 [DecidableEq α][Preorder α] : Multiset α → Prop := Acc MultisetRedLt
+
+/- MultisetRedLt is a special case of MultisetLT. -/
+theorem redLt_LT [DecidableEq α] [Preorder α] (M N : Multiset α) :
+    MultisetRedLt M N → MultisetLT M N := by
+  intro hyp
+  rcases hyp with ⟨X, Y, a, M_def, N_def, ys_lt_a⟩
+  apply MultisetLT.MLT Y {a} X _ M_def N_def
+  · simp; assumption
+  · simp
+
+/- Some useful lemmas about Multisets and the defined relations: -/
+lemma not_MultisetRedLt_0 [DecidableEq α] [LT α] (M: Multiset α) : ¬ MultisetRedLt M 0 := by
+  intro h
+  cases h with
+  | RedLt X Y a M nonsense _ =>
+    have contra : a ∈ (0 : Multiset α):= by
+      rw [nonsense]
+      simp_all only [Multiset.mem_add, Multiset.mem_singleton, or_true]
+    contradiction
+
+lemma mul_singleton_erase [DecidableEq α] {a : α} {M : Multiset α} :
+    M - {a} = M.erase a := by
+  ext
+  simp [Multiset.erase_singleton, Multiset.count_singleton]
+  split <;> simp_all
+
+lemma mul_mem_not_erase [DecidableEq α] {a a0: α} {M X : Multiset α}
+    (H : M = (a0 ::ₘ X).erase a) (hyp : ¬ a = a0) : a0 ∈ M := by
+  rw [H]
+  have : a0 ∈ (a0 ::ₘ X).erase a ↔ a0 ∈ (a0 ::ₘ X) := by
+    apply Multiset.mem_erase_of_ne
+    aesop
+  rw [this]
+  simp_all
+
+lemma mem_erase_cons [DecidableEq α] {a0: α} {M : Multiset α} (_ : a0 ∈ M) :
+    M = M - {a0} + {a0} := by
+  rw [add_comm]
+  simp_all [Multiset.singleton_add, mul_singleton_erase]
+
+lemma cons_erase [DecidableEq α] {a a0: α} {M X : Multiset α}
+    (H : a ::ₘ M = X + {a0}) : M = X + {a0} - {a} := by
+  if hyp : a = a0 then
```

Comment:
Please use `by_cases` instead of `if ... then`. This applies everywhere in the PR.